ghc-lib-gen/src/Ghclibgen.hs

```diff
@@ -295,10 +295,20 @@ calcParserModules ghcFlavor = do
       -- Lastly, manipulate text like 'GHC/Exts/Heap/Constants.hs'
       -- into 'GHC.Exts.Heap.Constants'.
       modules = map (replace "/" "." . dropSuffix ".hs") strippedModulePaths
-  -- We put 'HeaderInfo' here because doing so means clients who
-  -- just want to do parsing don't need 'ghc-lib' (this module
-  -- is needed for parsing in the presence of dynamic pragmas).
-  return $ nubSort (modules ++ ["HeaderInfo"])
+      -- The modules in this list would by default end up in
+      -- ghc-lib. We intervene so that rather, they go into
+      -- ghc-lib-parser.
+      extraModules =
+        [ f | True -- Invariant over versions.
+            , f <- [ "HeaderInfo" ]
+        ] ++
+        [ f | ghcFlavor `elem` [ GhcMaster, Ghc8101 ] -- Modern GHC.
+            , f <- [ "GHC.Hs.Dump" ]
+        ] ++
+        [ f | ghcFlavor `notElem` [ GhcMaster, Ghc8101 ] -- Legacy GHC.
+            , f <- [ "HsDumpAst" ]
```

Comment:
I think rather than `elem` and `notElem` checks you are better off with just an `if` to make it clear that the second case hits when the first doesn’t.

Reply:
My idea was to build in support for extending those lists but as you think it's overkill, i will simplify